### ISSUE TITLE
Add support for generated columns (or column constraint escape hatch)

### ISSUE THREAD

**Issue by mallman** [groue/GRDB.swift#854]
Please provide support for specifying generated column constraints with `ColumnDefinition`. See https://sqlite.org/gencol.html.

Alternatively, provide an "escape hatch" on the `ColumnDefinition` interface allowing the manual specification of column constraints through a SQL string. See https://sqlite.org/syntax/column-constraint.html for the overall syntax support. Of course, SQL injection or general mistakes are always a concern. However, in this case we're talking about a schema definition, which is presumably tested and defined by a trusted party.

## Environment

**GRDB flavor(s):** Custom SQLite
**GRDB version:** 5.0.2
**Installation method:** manual Xcode build
**Xcode version:** 12.2 beta 2
**Swift version:** 5
**Platform(s) running GRDB:** macOS
**macOS version running Xcode:** Catalina 10.15.7

**Comment by groue** [comment]
This is a good idea, @mallman. Generated columns were shipped in SQLite 3.31.0. I was waiting for such a feature request.

Do you feel like contributing a pull request which adds this feature? We may start discussing the desired api.

**Comment by mallman** [comment]
I'm writing a PR right now. So far seems very straightforward. I am modeling the API after the current `ColumnDefinition` interface. We can review when I have something working.

**Comment by mallman** [comment]
I've gotten far enough in my patch that we should discuss some API and implementation details.

Column naming and usage in the codebase is complicated by generated columns. Based on the SQLite column definition grammar (where a generated column is defined as a class of column constraint, like `UNIQUE`, `CHECK` or `DEFAULT`), I suggest we should accept that the term `column` can refer to either a pre-3.31.0 or (_ordinary_) column or a generated column. When clarity is called for, we could call an ordinary column an `ordinaryColumn` in the codebase, but I don't think that term will be necessary. I suggest we pick one or two from `generatedColumn`, `genColumn` or `generatedCol` as a term in the codebase referring to the concept of a generated column.

I suggest we also introduce a more abstract term to refer to columns which are not insertable or updatable where we need not assume a generated column. There is more than one kind of non-writable column in SQLite. Another such case is a [hidden column](https://sqlite.org/vtab.html#hiddencol). I suggest we pick from a prefix of either `writable`, `insertable` or `updatable`, and that we use that like `{prefix}Col` or `{prefix}Column`. The SQLite codebase seems to be biased toward `insertable`—look for `COLFLAG_NOINSERT` in https://www.sqlite.org/cgi/src/file?name=src/sqliteInt.h&ci=fca8dc8b578f215a. I'll use that terminology in the rest of this comment until we've settled the matter for GRDB.

(Incidentally, this line of thought suggests we might introduce another name for a hidden column, such as `hiddenColumn`; and that we might introduce a more abstract term for a column that is selectable, which would refer to an ordinary, hidden or generated column. However, so far it looks like introducing these terms would be premature and unnecessary.)

That said, this does complicate the current use of `column` and `columns` terms in the codebase. The term `column` is inadequate in the current codebase when we work with a table with a generated column. In some cases, it should refer specifically to an insertable column. In others it can refer to any ordinary column. For example, suppose we have a table schema

```sql
CREATE TABLE test(
  a INTEGER PRIMARY KEY,
  b INT,
  c INT GENERATED ALWAYS AS (a*abs(b)) VIRTUAL
)
```

Suppose we define a codable, persistable record for `test` as

```swift
struct TestRecord: Codable, PersistableRecord, FetchableRecord, EncodableRecord {
    static let databaseTableName = "test"
    let a: Int
    let b: Int
    let c: Int
}
```

(I'm assuming this code is correct for the sake of example. Also, I may have thrown in more protocols than necessary.)

In this case, I've found [this use](https://github.com/groue/GRDB.swift/blob/v5.0.2/GRDB/Record/PersistableRecord.swift#L859) of `persistenceContainer.columns` to be incorrect. Fixing these problems will require some surgery in the codebase. As suggested above, I suggest we reuse the terms `column` and `columns` to refer to columns that are either ordinary or generated, and modify the codebase where necessary to use the terms `insertableColumn`, `insertableColumns`, `generatedColumn` or `generatedColumns` when we are in a context where insertable or generated columns are specifically required.

(Of course, this omits the details of how generated columns impact the Record API, but I think we can defer those.)

Keeping with this strategy, the [columns](https://github.com/groue/GRDB.swift/blob/v5.0.2/GRDB/Core/Database+Schema.swift#L332) function must be modified. This would be a breaking but necessary change for tables with generated columns. Currently it returns non-hidden rows only. It should be modified to return columns that are either ordinary or generated. This can be done with the query

```sql
SELECT * FROM pragma_table_xinfo(\(tableName.quotedDatabaseIdentifier)) WHERE hidden IN (0,2,3)
```

Note this query is somewhat fragile because we're referring to the magic numbers 2 and 3 in the `hidden` column. (These refer to virtual and stored generated columns, respectively.) I don't believe this is documented behavior. I take this as truth from the source code: https://www.sqlite.org/cgi/src/file?name=src/pragma.c&ci=fca8dc8b578f215a. (Look for `COLFLAG_NOINSERT`.) The correctness of this query can be verified with unit tests.

Then we can add a property to `ColumnInfo` such as `isGenerated` (or an optional enum property specifying the specific kind of generated column, and/or a required enum property specifying whether a column is ordinary or generated (or ordinary virtual generated or stored generated). We might also add a computed property `isInsertable`. Users of the `ColumnInfo` API can use one of these properties to filter the column list when necessary.

We also need to address the question of the API for defining generated columns with `ColumnDefinition`. Toward that end I've added a public function to that class with the interface

```swift
public func generatedAs(sql: String, _ qualification: GeneratedColumnQualification) -> Self
```

As you can see, I've added the concept of a _qualification_ to a generated column. The abstract concept of whether a generated column is virtual or stored does not seem to be specified by the [SQLite documentation](https://www.sqlite.org/gencol.html). I struggled to identify a good name for this, and "qualification" is the best I've come up with. I don't particularly like it. As an alternative, we can replace this parameter with something like `isVirtual: Bool`, equating `isVirtual: false` with a stored column. I'm more reluctant to take that approach. From a readability perspective, source code that uses `.virtual` or `.stored` looks more readable to me than `isVirtual: true` or `isVirtual: false`.

I've defined `GeneratedColumnQualification` as a member enum of `ColumnDefinition` like so

```swift
public enum GeneratedColumnQualification: String {
    case stored  = "STORED"
    case virtual = "VIRTUAL"
}
```

The raw values are used in the SQL generation in the implementation. As such `GeneratedColumnQualification` is a bit leaky. As an alternative, we could switch on `GeneratedColumnQualification` in the SQL generation to choose a string literal. That is probably the better thing to do from an API perspective.

We should also decide whether and how to use a pre-processor expression to provide or omit the use of the generated columns at various points of the API. It's not clear to me what that expression would be. I really haven't looked into it yet. I think it would have to be based on a minimum SQLite version of 3.31.0. There's probably something that does this.

There may be further ambiguities or implementation questions to address later, but this will take me pretty far. Thank you.

**Comment by groue** [comment]
Wow, thanks for the quality of your analysis, @mallman. Let's continue!

I skip your first paragraphs because they do not need any specific answer. They provide a great setup for this conversation.

> Suppose we define a codable, persistable record for test as [...] In this case, I've found [this use](https://github.com/groue/GRDB.swift/blob/v5.0.2/GRDB/Record/PersistableRecord.swift#L859) of persistenceContainer.columns to be incorrect.

We could filter generated columns from the generated UPDATE and INSERT statements. But do we need it?

It seems to me that defining a persistable record which has properties that map generated columns is begging for trouble. Inserts and updates of those columns is meaningless, and this questions the mere presence of those properties in a persistable record. Besides, every time you instantiate or modify such a record, it is hard to guarantee that the Swift code keeps generated properties synchronized with generated columns. Talk about [Object–relational impedance mismatch](https://en.wikipedia.org/wiki/Object–relational_impedance_mismatch)!

Let's take an example: a `player` table has a generated `totalScore = score + bonus` colum. Can we design a safe Swift api which prevents the app from shooting itself in the foot? It could be something like:

```swift
struct Player: Decodable, FetchableRecord, TableRecord {
    // `let` properties for high fidelity with the database.
    let id: Int64
    let score: Int
    let bonus: Int
    let totalScore: Int
    
    // Extra safety: disable all initializers but `init(row:)`
    @available(*, unavailable) init() { fatalError() }
}

struct PersistedPlayer: Encodable, MutablePersistableRecord {
    static let databaseTableName = Player.databaseTableName
    var id: Int64?
    var score: Int
    var bonus: Int
    
    // Update auto-incremented id upon successful insertion
    mutating func didInsert(with rowID: Int64, for column: String?) {
        id = rowID
    }
}

// Convenience methods needed by the application
extension PersistedPlayer {
    mutating func insertAndFetchPlayer(_ db: Database) throws -> Player {
        assert(db.isInsideTransaction) // Paranoid. A savepoint is another way to deal with it.
        try insert(db)
        return try Player.fetchOne(db, key: id!)!
    }
}

// Usage
try dbQueue.write { db in
    var persistedPlayer = PersistedPlayer(id: nil, score: 1000, bonus: 100)
    let player = try persistedPlayer.insertAndFetchPlayer(db)
    assert(player.totalScore == 1100)
}
```

If such a setup is fostered by the documentation, then we do not need to filter generated columns from inserts and updates, to do anything to the record apis, or to alter `Database.columns(in:)` with support for `pragma_table_xinfo`. We just don't have to care about generated columns beyond the table creation and alteration apis (addressed below).

Now, since this issue is your feature request, you may have another vision of how you intend to use the feature. I'm all listening!

> We also need to address the question of the API for defining generated columns with `ColumnDefinition`. [...]

I evaluated a distinct `GeneratedColumnDefinition` type that would prevent the user from defining invalid generated columns (by removing the `primaryKey()` and `defaults(to:)` methods). But I'm pleased by your suggestion. Mistakes are rare and trigger an SQLite error. `generatedAs` composes well with other column methods. I'm all with you for avoiding booleans for distinguishing `virtual` from `stored`. PostgreSQL, the major inspiration source for SQLite, does not support other kinds of generated columns, but it is obvious that they designed an extensible syntax.

```swift
// LGTM
try db.create(table: "player") { t in
    t.column("score", .integer)
    t.column("bonus", .integer)
    t.column("totalScore").generatedAs(sql: "score + bonus")
    t.column("totalScore").generatedAs(sql: "score + bonus", .stored)
    t.column("totalScore")
        .generatedAs(sql: "score + bonus", .stored)
        .notNull()
        .indexed()
        .check { $0 >= 0 }
}
```

**Temporary conclusion**

A pull request that "just" adds `generatedAs(sql:_:)` looks quite sufficient to me, in table creation, and table alteration. That would be a great contribution!

You asked about SQL injection, and there a few paragraphs about it below.

You also asked for a general escape hatch, and I agree that we miss it. Let's make it the topic of another issue, and keep this one focused on a specific Swift api for generated columns.

```swift
try db.create(table: "player") { t in
    t.column("score", .integer)
    t.column("bonus", .integer)
    // escape hatch
    t.column("totalScore INT GENERATED ALWAYS AS (score + bonus) VIRTUAL")
}
```

And I'll submit a pull request to http://github.com/swiftlyfalling/SQLiteLib for SQLite 3.33.0 :-)

What do you think, @mallman?

<details>
    <summary>A note about SQL injection</summary>

One can only perform SQL injection with GRDB by providing a raw SQL string, as below. Raw SQL strings are completely beyond our control, and we do not have to care about SQL injection with them:

```swift
// ¯\_(ツ)_/¯
let value: String = ...
try db.execute(sql: "INSERT ... '\(value)' ...")
```

GRDB has two APIs that avoid SQL injection:

- Query interface expressions. For example:

    ```swift
    try db.create(table: "player") { t in
        t.column("name").check { $0 != "O'Brien" }
        t.column("score", .integer)
        t.column("bonus", .integer)
        t.column("totalScore").generatedAs(Column("score") + Column("bonus"))
    }
    ```
    
    I think it is desirable to support query interface expressions for defining generated columns. I'll add them if you do not. If you intend to add support for query interface expressions, here is some guidance:
    
    Those expressions are based on the `SQLExpression` protocol. It comes with the [`quotedSQL(_:)`](https://github.com/groue/GRDB.swift/blob/v5.0.2/GRDB/QueryInterface/SQL/SQLExpression+QueryInterface.swift#L7) method which makes sure all literals are properly quoted. You'll see it used in TableDefinition.swift, because SQLite won't accept query arguments (`?`) in table creation. The escaping technique itself is based on the built-in [quote(X)](https://sqlite.org/lang_corefunc.html#quote) SQLite function, which means that it's as reliable as possible.

- [SQL interpolation](https://github.com/groue/GRDB.swift/blob/master/Documentation/SQLInterpolation.md). For example:

    ```swift
    // safe
    let value: String = ...
    try db.execute(literal: "INSERT ... \(value) ...")
    ```
    
    SQL interpolation is not available *everywhere* in GRDB apis, and is currently totally absent from the table creation apis. I'm not sure we need it for generated columns.
    
    After all, dangerous literals are not very common there (don't you think?). Besides, SQL interpolation is not expected by most users. I have a hard time figuring people actively looking for this feature (and being disappointed it is missing). I have always favored query interface expressions instead, both in apis and in documentation.

</details>

**Comment by mallman** [comment]
Hi @groue. I spent some time rethinking my approach based on your comments.

I agree that generated columns call for a more nuanced approach to records. In my case, I was using the same struct for fetching and persisting. I included a constructor argument for the generated column value, and provided it manually when necessary.

The approach of separating the record into fetchable and persistable pairs is going better. However, I'm running into a problem with associations. GRDB is misaligning table column boundaries when a table with a generated column is involved. I believe the problem is with how it computes the number of columns in a `SELECT *`. I'm referring to the code at https://github.com/groue/GRDB.swift/blob/v5.0.2/GRDB/QueryInterface/SQL/SQLSelectable%2BQueryInterface.swift#L77.

The `table_info` pragma does not return generated columns. If I modify the `Database.columns` function to

```sql
SELECT * FROM pragma_table_xinfo('\(tableName)') WHERE hidden IN (0,2,3)
```

then all of my application code appears to work. Additionally, all of the GRDB unit tests pass except for tests in [AssociationPrefetchingSQLTests.swift](https://github.com/groue/GRDB.swift/blob/v5.0.2/Tests/GRDBTests/AssociationPrefetchingSQLTests.swift). These fail because they assume that the query string at https://github.com/groue/GRDB.swift/blob/v5.0.2/GRDB/Core/Database+Schema.swift#L369 does not include the substring `"SELECT"`.

Anyway, I'll submit my PR as-is and we can pick up the discussion there.

**Comment by mallman** [comment]
PR is https://github.com/groue/GRDB.swift/pull/855.

**Comment by groue** [comment]
> However, I'm running into a problem with associations.

Well spotted!

I join you in #855.

**Comment by groue** [comment]
Escape hatch was implemented in 085c8f59. It will ship in next version:

> Columns can also be defined with a raw sql String, or an [SQLLiteral](https://github.com/groue/GRDB.swift/blob/master/Documentation/SQLInterpolation.md#sqlliteral) in which can safely embed raw values without any risk of syntax errors or SQL injection:
>
> ```swift
> try db.create(table: "player") { t in
>     t.column(sql: "name TEXT")
>    
>     let defaultName: String = ...
>     t.column(literal: "name TEXT DEFAULT \(defaultName)")
> }
> ```
 
There is also `t.constraint(sql:)` and `t.constraint(literal:)` for adding a table constraint.
